plugin/call_hierarchy.py

```diff
@@ -0,0 +1,157 @@
+from .core.protocol import CallHierarchyIncomingCallsParams
+from .core.protocol import CallHierarchyItem
+from .core.protocol import CallHierarchyOutgoingCall
+from .core.protocol import CallHierarchyOutgoingCallsParams
+from .core.protocol import CallHierarchyPrepareParams
+from .core.protocol import Request
+from .core.registry import new_tree_view_sheet
+from .core.registry import windows
+from .core.registry import get_position
+from .core.registry import LspTextCommand
+from .core.registry import LspWindowCommand
+from .core.tree_view import TreeDataProvider
+from .core.tree_view import TreeItem
+from .core.typing import cast
+from .core.typing import IntEnum, List, Optional
+from .core.views import parse_uri
+from .core.views import SYMBOL_KINDS
+from .core.views import text_document_position_params
+from functools import partial
+import sublime
+
+
+class CallHierarchyType(IntEnum):
+    IncomingCalls = 1
+    OutgoingCalls = 2
+
+
+class CallHierarchyDataProvider(TreeDataProvider):
+
+    def __init__(
+        self,
+        window: sublime.Window,
+        session_name: str,
+        call_hierarchy_type: CallHierarchyType,
+        root_elements: List[CallHierarchyItem]
+    ) -> None:
+        self.window = window
+        self.session_name = session_name
+        self.call_hierarchy_type = call_hierarchy_type
+        self.root_elements = root_elements
+
+    def get_children(self, element: Optional[CallHierarchyItem]) -> Promise[List[CallHierarchyItem]]:
+        if element is None:
+            return Promise.resolve(self.root_elements)
+        wm = windows.lookup(self.window)
+        if not wm:
+            return Promise.resolve([])
+        for session in wm.get_sessions():
+            if not session.has_capability('callHierarchyProvider'):
+                continue
+            if session.config.name == self.session_name:
+                break
+        else:
+            return Promise.resolve([])
+        if self.call_hierarchy_type == CallHierarchyType.IncomingCalls:
+            params = cast(CallHierarchyIncomingCallsParams, {'item': element})
+            return session.send_request_task(Request.incomingCalls(params)).then(self._handle_incoming_calls_async)
+        elif self.call_hierarchy_type == CallHierarchyType.OutgoingCalls:
+            params = cast(CallHierarchyOutgoingCallsParams, {'item': element})
+            return session.send_request_task(Request.outgoingCalls(params)).then(self._handle_outgoing_calls_async)
+        return Promise.resolve([])
+
+    def get_tree_item(self, element: CallHierarchyItem) -> TreeItem:
+        command_url = sublime.command_url('lsp_open_location', {
+            'location': {
+                'targetUri': element['uri'],
+                'targetRange': element['range'],
+                'targetSelectionRange': element['selectionRange']
+            },
+            'session_name': self.session_name,
+            'flags': sublime.SEMI_TRANSIENT | sublime.ADD_TO_SELECTION
```

Comment:
```suggestion
            'flags': sublime.ADD_TO_SELECTION | sublime.CLEAR_TO_RIGHT
```

I would add  `sublime.CLEAR_TO_RIGHT`, to avoid too many open views.
![Screenshot 2022-12-28 at 11 37 37](https://user-images.githubusercontent.com/115091796/209799279-d20f5642-e597-4d25-a756-678cd2e48b43.png)

Reply:
I see the exact behavior from your screenshot with `sublime.ADD_TO_SELECTION | sublime.CLEAR_TO_RIGHT`.
But I see now that even without CLEAR_TO_RIGHT this still happens if the links are to different files.
It's probably a bug which has been fixed on the latest dev built. So maybe it's better to remove the `"flags"` argument alltogether, until there is another ST stable build with the bugfix?

Reply:
I use another approach now:
If you hold <kbd>Ctrl</kbd> or <kbd>Shift</kbd> when clicking on the link, then it will be added to the selection, and otherwise it will open as a single regular tab which takes the focus. This means that you have to know that you have to use the modifier key if you want the call hierarchy tab to be still visible after clicking on a item label, but this is consistent with how the built-in definitions/references popup from ST works.

We can reconsider using the ADD_TO_SELECTION flag by default when a new stable build of ST with the mentioned bugfix lands.

Reply:
I can confirm that `sublime.CLEAR_TO_RIGHT` works nice in the latest dev build.